src/lib/sql/sql_pipeline.hpp

```diff
@@ -72,11 +74,14 @@ class SQLPipeline : public Noncopyable {
   // Returns the entire execution time
   std::chrono::microseconds execution_time_microseconds();
 
+  void set_optimizer(const std::shared_ptr<Optimizer>& optimizer);
+
  private:
-  SQLPipeline(const std::string& sql, std::shared_ptr<TransactionContext> transaction_context, bool use_mvcc);
+  SQLPipeline(const std::string& sql, std::shared_ptr<TransactionContext> transaction_context, ChunkUseMvcc use_mvcc);
 
   std::vector<std::shared_ptr<SQLPipelineStatement>> _sql_pipeline_statements;
-  size_t _num_statements;
+
+  std::shared_ptr<SQLPipelineControlBlock> _control_block;
```

Comment:
This feels really weird. It encapsulates things that belong together only loosely solely for the purpose of passing them in one piece.

Reply:
To me this doesn't feel weird at all, but how you should do it! Previously, we had state that had to be the same for all Statements but was stored in redundantly. Now, it is shared state that is guaranteed to not diverge. How is this weird?

Reply:
Does not diverge as in `_transaction_context(control_block->transaction_context),`? ;)

It's more that I find it weird to throw almost unrelated things together in one struct.

Reply:
Well spotted, I don't like that aspect either... (But don't agree on the "unrelated things" part - they are related, they are shared state of pipeline and statements.)

Braindump:

From an interface perspective, there are three modes regarding mvcc:
- Disable
- Enabled, with an explicit TransactionContext being passed in by the user, which is the same for all Statements
- Enabled, with the each Statements creating its own TransactionContext in the background.

With the `auto_commit` field gone, the state shared by the Pipeline/Statements is the Optimizer, which one of these three modes is chosen and optionally the global TransactionContext. Really want the Optimizer to be shared among the Pipeline/Statements in a way that changing the variable it is stored in is guaranteed to affect all Statements.

Reply:
I mean, I could just have a `std::shared_ptr<std::shared_ptr<Optimizer>>` field in Statements and Pipeline and leave the TransationContext sharing as it was.

Reply:
Why would you even change the optimizer of an existing pipeline? If the optimized LQP is already generated, it is cached and your change does not have an effect. Shouldn't the optimizer be set when you create the pipeline?

Reply:
Yeah, wanted to somewhat restrict the number of constructor args, but you're right.